### ISSUE TITLE
DS communication issue when using NCCL backend: All_reduce instead of reduce_scatter (or several reduce ops)

### ISSUE THREAD

**Issue by aperezdieguez** [microsoft/DeepSpeed#5758]
I am executing DS for multi-node training. I am using the pdsh launcher and the NCCL backend from PyTorch DDP. I am cloning the repo and installing DS via pip. 

When I execute DS with either ZeRO Stage 1 or Stage 2, I found that it is always performing an All_reduce collective, when the original paper suggests Reduce_scatter.  I have the following DS zero optimization configuration:

`
zero_optimization": {
        "contiguous_gradients": true, 
        "gather_16bit_weights_on_model_save": true, 
        "overlap_comm": true, 
        "reduce_scatter": true, 
        "stage": 2
`

Based on #264, I should observe that the gradient calculation is performed with a set of Reduce operations in Stage 2, which implement a Reduce_scatter pattern, when using that configuration. Instead, I observe AllReduce collectives happening in both Stage 1 and Stage 2 executions when I use `NCCL_DEBUG=TRACE` .  Is this a bug?

**Comment by GuanhuaWang** [comment]
Hi @aperezdieguez , 

thx for trying reduce scatter in z1 and z2. It is true that z1/z2 always use allreduce for gradient reduction regardless of whether you set self.reduce_scatter flag to true or false. As you can search `dist.all_reduce` in this implementation file https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage_1_and_2.py and there is no `reduce_scatter` function calls.

**Comment by aperezdieguez** [comment]
@GuanhuaWang so why there is not a `reduce_scatter` collective in the current implementation as the original paper suggests? shouldn't this be a performance optimization? Older versions seemed to implement the `reduce_scatter` collective as a cascade of  `reduce` operations in stage 2, based on your older posts (#264) Any clarification will be appreciated, thank you!

**Comment by GuanhuaWang** [comment]
> @GuanhuaWang so why there is not a `reduce_scatter` collective in the current implementation as the original paper suggests? shouldn't this be a performance optimization? Older versions seemed to implement the `reduce_scatter` collective as a cascade of `reduce` operations in stage 2, based on your older posts (#264) Any clarification will be appreciated, thank you!

Hi @aperezdieguez ,

thx for the question. 

Since Zero stage 1 and 2, we never partition weights among GPUs, thus each GPU need to use **Whole** gradients to updates weights. Therefore, we **don't need to partition (i.e. scatter)** aggregated gradients after reduction, thus allreduce is more preferable.

**Comment by aperezdieguez** [comment]
@GuanhuaWang  thanks for your prompt answer.

Could you clarify whether this was always the case or if earlier there was another implementation? Maybe I am understanding wrongly, but this seems to be opposite to what paper says: 

> 5.2 Gradient Paritioning (Page 10)
> Therefore, as each gradient of each
layer becomes available during the backward propagation, we only reduce them on the data
parallel process responsible for updating the corresponding parameters [..]vEffectively this is a Reduce-Scatter operation, where gradients corresponding to different parameters are reduced to different process

and also:

> 7.2.1 Communication Volume with P_os+g (Page 13)
> With gradient partitioning, each process only stores the portion of the gradients, that is required
to update its corresponding parameter partition. As such, instead of an all-reduce, ZeRO only
requires a scatter-reduce operation on the gradients.

Although you never partition weights among GPUs, it seems that each GPU is only updating the weights associated to the partitioned optimizer states assigned to that GPU (as explained in the paper). Considering Stage 2 where gradients are distributed among GPU, each GPU only requires the corresponding reduced gradients for its states, not the reduced gradients for all weights. Otherwise, the  communication volumes claimed in the paper for Stage 2 - says that St2 communicates 2 times the number of parameters in the model- wouldn't be true.  What am I missing? Thank you,

**Comment by GuanhuaWang** [comment]
> @GuanhuaWang thanks for your prompt answer.
> 
> Could you clarify whether this was always the case or if earlier there was another implementation? Maybe I am understanding wrongly, but this seems to be opposite to what paper says:
> 
> > 5.2 Gradient Partitioning (Page 10)
> > Therefore, as each gradient of each
> > layer becomes available during the backward propagation, we only reduce them on the data
> > parallel process responsible for updating the corresponding parameters [..] Effectively this is a Reduce-Scatter operation, where gradients corresponding to different parameters are reduced to different process
> 
> and also:
> 
> > 7.2.1 Communication Volume with P_os+g (Page 13)
> > With gradient partitioning, each process only stores the portion of the gradients, that is required
> > to update its corresponding parameter partition. As such, instead of an all-reduce, ZeRO only
> > requires a scatter-reduce operation on the gradients.
> 
> Although you never partition weights among GPUs, it seems that each GPU is only updating the weights associated to the partitioned optimizer states assigned to that GPU (as explained in the paper). Considering Stage 2 where gradients are distributed among GPUs, each GPU only requires the corresponding reduced gradients for its states, not the reduced gradients for all weights. Otherwise, the communication volumes claimed in the paper for Stage 2 - says that St2 communicates 2 times the number of parameters in the model- wouldn't be true. What am I missing? Thank you,

Hi @aperezdieguez , thx for the thoughts here. After going through the code thoroughly, I think my previous explanation is wrong for `reduce_scatter=True` case of zero stage 1 and 2. 

Basically, if you want to observe gradient calculation is performed with a set of Reduce() when `reduce_scatter=True`, you should set `use_multi_rank_bucket_allreduce` as `False`.  More details is illustrated as below:

[average tensor() ](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage_1_and_2.py#L1041) is the entrance function we used for gradient reduction. At the very end, we have a loop with `use_multi_rank_bucket_allreduce` condition:

```
                if self.use_multi_rank_bucket_allreduce:
                    self.allreduce_and_scatter(buckets[bucket_key],
                                               numel_per_bucket=self.reduce_bucket_size,
                                               divide=False,
                                               process_group=bucket_key)
                else:
                    dst, process_group = bucket_key
                    self.allreduce_no_retain(buckets[bucket_key],
```

As above conditions, if `use_multi_rank_bucket_allreduce =True`, then it will call `AllReduce` on every gradient slice, else it will call `Reduce` and put aggregated gradient piece onto destination GPU.
